primitives/tracing/src/span_dispatch.rs

```diff
@@ -0,0 +1,49 @@
+use tracing::{span, Level};
+
+macro_rules! span_dispatch {
+	($target:tt, $name:tt, { $($registered_target:tt,$registered_name:tt;)* }) => {
+		match ($target, $name) {
+			$(
+				($registered_target, $registered_name) => Ok(span!(target: $registered_target, Level::INFO, $registered_name, is_valid_trace = true)),
+			)*
+			_ => {
+				Err(format!("Trying to profile span target: {}, name: {} that is not registered",
+					$target,
+					$name))
+			}
+		}
+	}
+}
+
+/// Spans (target/name pairs) must be pre-registered here.
+/// To allow tracing with earlier on-chain versions of the runtime,
+/// it is necessary to keep target/name pairs, even if not used in the current runtime.
```

Comment:
I'm not sure I understand why it has to be pre-regsitered, can you elaborate?

Reply:
Ah, so you mean it simply has to be a string literal that is probably consumed during compilation not `&'static str`. I think I do understand now.

To avoid that I guess we could either:
1. Have a build script to generate this mappings by looking into usages.
2. Alternatively to make sure that whoever uses that in the runtime updates this part as well, we could only allow using it via some constant like so:
```rust
pub struct TracingSpan(&'static str);
const FRAME_EXECUTIVE = TracingSpan("frame_executive");

pub fn create_registered_span(target: &TracingSpan, name: &TracingSpan) -> Result<tracing::Span, String> {
	span_dispatch! {
		target.0, name.0, {
			"frame_executive","execute_block_wasm";
                         ...
```
But IMHO both of these could be done as a follow-up instead of being part of this PR.

Reply:
That's a neat idea to use a build script - if we have to stick with this method then probably the best option. I'm actually going to investigate a 3rd possibility - to create all spans from wasm using the same target/name identifier (maybe target: `wasm_trace` name: `wasm_trace`) and then use associated values to record the actual target and name, which could be handled in our global subscriber - not strictly conforming to the design intent of the library, but if it works it could be the cleanest solution.